### PR TITLE
Fix the ProneOffsets of TakeCover

### PR DESCRIPTION
**Pull request by abcdefg30**
On current bleed the bullets of prone infantry are instantly blocked. Also see #12683.

### CODE REVIEW

File: mods/cnc/rules/infantry.yaml [hunk 1]
```diff
@@ -125,6 +131,8 @@ E5:
 		LocalOffset: 341,0,256
 		FireDelay: 3
 		MuzzleSequence: muzzle
+	TakeCover:
+		ProneOffset: 400,0,-256
```
Comment (pchote) [inline mods/cnc/rules/infantry.yaml]: This looks quite wrong ingame.

File: mods/cnc/rules/infantry.yaml [hunk 2]
```diff
@@ -94,6 +98,8 @@ E4:
...
```
Comment (pchote) [inline mods/cnc/rules/infantry.yaml]: This looks quite wrong ingame.

File: mods/d2k/rules/infantry.yaml [hunk 3]
```diff
@@ -175,6 +177,8 @@ grenadier:
 		Weapon: grenade
 		LocalOffset: 192,0,224
 		FireDelay: 3
+	TakeCover:
+		ProneOffset: 300,0,-224
```
Comment (pchote) [inline mods/d2k/rules/infantry.yaml]: This looks quite wrong ingame.

File: mods/ra/rules/infantry.yaml [hunk 4]
```diff
@@ -507,6 +513,8 @@ SHOK:
 	Armament@GARRISONED:
 		Name: garrisoned
 		Weapon: PortaTesla
+	TakeCover:
+		ProneOffset: 500,0,-341
```
Comment (pchote) [inline mods/ra/rules/infantry.yaml]: This is too far forward too.  Make sure that you use the combat debug overlay and compare the muzzle position before and after going prone.

File: mods/ra/rules/infantry.yaml [hunk 5]
```diff
@@ -158,6 +162,8 @@ E4:
 	Armament@GARRISONED:
 		Name: garrisoned
 		Weapon: Flamer
+	TakeCover:
+		ProneOffset: 500,0,-341
```
Comment (pchote) [inline mods/ra/rules/infantry.yaml]: This is quite wrong, and I just noticed that the standing fire offsets are wrong too:

screen shot 2017-02-04 at 11 19 11
"It burns when I pee!"

File: mods/ra/rules/infantry.yaml [hunk 6]
```diff
@@ -96,6 +96,8 @@ E2:
 		Name: garrisoned
 		Weapon: Grenade
 		FireDelay: 15
+	TakeCover:
+		ProneOffset: 0,0,-555
```
Comment (pchote) [inline mods/ra/rules/infantry.yaml]: This is off too.

**Review by pchote — CHANGES_REQUESTED** [review]
There are quite a few issues with the new offsets in the shipping mods.  I haven't checked TS.

**Comment by reaperrr** [comment]
I'm beginning to wonder whether we should add a boolean to completely disable the terrain height check or a lenience threshold to `Bullet` afterall, as this feature right now it seems to only increase regression risk for 2D mods.

**Comment by pchote** [comment]
It is only regressing things that were already bogus, so while it is painful for everyone my preference is to fix the bogus things.

**Comment by abcdefg30** [comment]
While I often don't agree with your nits, I tried to adjust the values even further and hope it's ok for you now.

**Comment by pchote** [comment]
We have a visualization specifically for setting up these firing offsets, so I don't think there is a good excuse for not having these set up properly.

Consider the d2k grenadier:
grenadier

The square end of the indicator should line up with the end of the barrel (in this case the position of the infantry's hand at the end of the throw).  The low resolution makes it kind of hard to guess, but the current position in this PR (white indicator) is quite clearly off from my best guess (red indicator).

I'd say that the current position is about 2px too far from the center of the actor, and about 4px too low.  D2k uses 32 WDist per pixel, so that means that the `x` offset needs to be reduced by ~64 and the `z` offset increased by ~128.  Tweaking the offset to `128,0,-96` does indeed look much better ingame.  There is an ambiguity between the `y` and `z` offsets when looking to the sides and the `x` and `z` offsets when looking north/south, but that can be resolved by checking the infantry from the four main facings.

**Comment by GraionDilach** [comment]
>We have a visualization specifically for setting up these firing offsets

I wouldn't call it that way. Yes, it shows the current used offsets, but you still have to guesswork the actual values AND you have to restart the game for every guess. If DebugCombatOverlay would have an interface to change the currently selected unit's firing indicator position with dev commands alone, then I'd say that is correct.

**Comment by pchote** [comment]
Sure, but it's more precise than just guesswork: the relationship between WDist and screen pixels are known, so it shouldn't take more than a couple of iterations to set up each actor if you're able to visualize the multiple 2d views -> 3d coordinates in your head.

**Comment by GraionDilach** [comment]
When the modder ends up with a loading time over a minute, turret offsets, different voxel scales (not talking about RenderVoxels here but the proper Scale in voxel header) even those couple of iterations would take a good while. The relationship between WDist and screen pixels are only known easily in gen1 mods, it's more complex in isometric.

**Comment by pchote** [comment]
That's better discussed in its own issue, because this PR doesn't involve long loading times, turret offsets, or voxel scales.

**Comment by abcdefg30** [comment]
Rebased and updated.

**Review by pchote — APPROVED** [review]

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/c1197e2c261d84b4d85f9f3d408107d6e824d31f...2774d81adc66ad2df1e1f7d81ed25313405169e1)

**Pull request merged**